src/Http/Routing/src/Matching/DfaMatcher.cs

```diff
@@ -10,6 +11,8 @@ namespace Microsoft.AspNetCore.Routing.Matching;
 
 internal sealed partial class DfaMatcher : Matcher
 {
+    private const int CandidateSetStackSize = 4;
```

Comment:
is 4 just from heuristic/observation or is there some deeper reason

Reply:
Heuristic, usually there's a single match per route, more complex MVC/WebAPI applications might have policies and other hooks that avoid this fast path.